### ISSUE TITLE
Variables moved in match guards

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#1408]
https://github.com/rust-lang/rust/issues/29723 is pretty scary to me:

```rust
fn main() {
    let s = String::from("unsound");
    match 0 {
        0 if { drop(s); false } => {}
        _ => println!("{:?}", s) // output: "°Ґ\u{7}V\u{0}"
    }
}
```

Is there anything Clippy can do to help catch this? Even if there are false positives, I would love to run `cargo clippy -- -D move_in_match_guard` against all of my projects just to find the places I need to audit.

**Comment by flip1995** [comment]
In 1.36 and 1.37 this was only a warning. From 1.38 this is a hard error. Thanks for noticing!
